components/script/dom/document.rs

```diff
@@ -1368,6 +1370,84 @@ impl Document {
             self.commit_focus_transaction(FocusType::Element, can_gc);
             self.maybe_fire_dblclick(client_point, node, pressed_mouse_buttons, can_gc);
         }
+
+        // When the contextmenu event is triggered by right mouse button
+        // the contextmenu event MUST be dispatched after the mousedown event.
+        if let (MouseEventType::MouseDown, MouseButton::Right) = (mouse_event_type, button) {
+            self.maybe_show_context_menu(
+                node.upcast(),
+                pressed_mouse_buttons,
+                client_point,
+                can_gc,
+            );
+        }
+    }
+
+    /// <https://www.w3.org/TR/uievents/#maybe-show-context-menu>
+    fn maybe_show_context_menu(
+        &self,
+        target: &EventTarget,
+        pressed_mouse_buttons: u16,
+        client_point: Point2D<f32>,
+        can_gc: CanGc,
+    ) {
+        let client_x = client_point.x.to_i32().unwrap_or(0);
+        let client_y = client_point.y.to_i32().unwrap_or(0);
+
+        // <https://w3c.github.io/uievents/#contextmenu>
+        let menu_event = PointerEvent::new(
+            &self.window,
+            None,
+            DOMString::from("contextmenu"),
+            EventBubbles::Bubbles,
+            EventCancelable::Cancelable,
+            Some(&self.window),
+            0,
+            client_x,
+            client_y,
+            client_x,
+            client_y,
+            false,
+            false,
+            false,
+            false,
+            2i16, // right mouse button
+            pressed_mouse_buttons,
+            None,
+            None,
+            // TODO: decide generic pointer id
+            // <https://www.w3.org/TR/pointerevents3/#dom-pointerevent-pointerid>
+            0,
+            1,
+            1,
+            0.5,
+            0.0,
+            0,
+            0,
+            0,
+            PI / 2.0,
+            0.0,
+            DOMString::from("mouse"),
+            true,
+            vec![],
+            vec![],
+            can_gc,
+        );
```

Comment:
Wow. This is a massive data structure! I think at the very least we need to have the names of the parameters here as comments.

Reply:
updated!